modules/core/locations.js

```diff
@@ -0,0 +1,267 @@
+import LocationConflation from '@ideditor/location-conflation';
+import whichPolygon from 'which-polygon';
+import calcArea from '@mapbox/geojson-area';
+import { utilArrayChunk } from '../util';
+
+let _mainLocations = coreLocations(); // singleton
+export { _mainLocations as locationManager };
+
+//
+// `coreLocations` maintains an internal index of all the boundaries/geofences used by iD.
+// It's used by presets, community index, background imagery, to know where in the world these things are valid.
+// These geofences should be defined by `locationSet` objects:
+//
+// let locationSet = {
+//   include: [ Array of locations ],
+//   exclude: [ Array of locations ]
+// };
+//
+// For more info see the location-conflation and country-coder projects, see:
+// https://github.com/ideditor/location-conflation
+// https://github.com/ideditor/country-coder
+//
+export function coreLocations() {
+  let _this = {};
+  let _resolvedFeatures = {};              // cache of *resolved* locationSet features
+  let _loco = new LocationConflation();    // instance of a location-conflation resolver
+  let _wp;                                 // instance of a which-polygon index
+
+  // pre-resolve the worldwide locationSet
+  const world = { locationSet: { include: ['Q2'] } };
+  resolveLocationSet(world);
+  rebuildIndex();
+
+  let _queue = [];
+  let _deferred = new Set();
+  let _inProcess;
+
+
+  // Returns a Promise to process the queue
+  function processQueue() {
+    if (!_queue.length) return Promise.resolve();
+
+    // console.log(`queue length ${_queue.length}`);
+    const chunk = _queue.pop();
+    return new Promise(resolvePromise => {
+        const handle = window.requestIdleCallback(() => {
+          _deferred.delete(handle);
+          // const t0 = performance.now();
+          chunk.forEach(resolveLocationSet);
+          // const t1 = performance.now();
+          // console.log('chunk processed in ' + (t1 - t0) + ' ms');
+          resolvePromise();
+        });
+        _deferred.add(handle);
+      })
+      .then(() => processQueue());
+  }
+
+  // Pass an Object with a `locationSet` property,
+  // Performs the locationSet resolution, caches the result, and sets a `locationSetID` property on the object.
+  function resolveLocationSet(obj) {
+    if (obj.locationSetID) return;  // work was done already
+
+    try {
+      const locationSet = obj.locationSet;
+      if (!locationSet) {
+        throw new Error('object missing locationSet property');
+      }
+      const resolved = _loco.resolveLocationSet(locationSet);
+      const locationSetID = resolved.id;
+      obj.locationSetID = locationSetID;
+
+      if (!resolved.feature.geometry.coordinates.length || !resolved.feature.properties.area) {
+        throw new Error(`locationSet ${locationSetID} resolves to an empty feature.`);
+      }
+      if (!_resolvedFeatures[locationSetID]) {  // First time seeing this locationSet feature
```

Comment:
Good question, and I don't really remember why I wrote it this way.  
I think it might just have been defensive coding - this code makes a copy of the Object in `_resolvedFeatures` for whatever iD wants to do with it, instead of passing around references to the `_loco` internally cached Objects.

Reply:
Got it. no-op for now as impact is low and priority is pushing this change through